src/clarity/database/datastore.rs

```diff
@@ -7,35 +9,65 @@ use crate::clarity::types::QualifiedContractIdentifier;
 use crate::clarity::util::hash::Sha512Trunc256Sum;
 use crate::clarity::StacksBlockId;
 use std::collections::HashMap;
+use std::hash::Hash;
 
 #[derive(Clone, Debug)]
 pub struct Datastore {
-    store: HashMap<String, String>,
+    store: HashMap<StacksBlockId, HashMap<String, String>>,
     metadata: HashMap<(String, String), String>,
-    chain_tip: StacksBlockId,
+    open_chain_tip: StacksBlockId,
+    current_chain_tip: StacksBlockId,
     chain_height: u32,
+    height_at_chain_tip: HashMap<StacksBlockId, u32>,
```

Comment:
could specify which chain tip in the variable name here 

so either `height_at_open_chain_tip` or `height_at_current_chain_tip`

Reply:
not necessary though, I think most people would assume it is current chain tip